chia/wallet/conditions.py

```diff
@@ -0,0 +1,1088 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from dataclasses import dataclass, replace
+from typing import Any, Dict, Iterable, List, Optional, Type, TypeVar, Union
+
+from blspy import G1Element
+from clvm.casts import int_to_bytes
+
+from chia.types.blockchain_format.program import Program
+from chia.types.blockchain_format.sized_bytes import bytes32
+from chia.types.condition_opcodes import ConditionOpcode
+from chia.util.hash import std_hash
+from chia.util.ints import uint32, uint64
+from chia.util.streamable import Streamable, streamable
+
+_T_ConditionSubclass = TypeVar("_T_ConditionSubclass")
+
+
+class Condition(Streamable, ABC):
+    @abstractmethod
+    def to_program(self) -> Program:
+        ...
+
+    @classmethod
+    @abstractmethod
+    def from_program(cls: Type[_T_ConditionSubclass], program: Program, **kwargs: Any) -> _T_ConditionSubclass:
+        ...
+
+
+@streamable
+@dataclass(frozen=True)
+class AggSigParent(Condition):
+    pubkey: G1Element
+    msg: bytes
+    parent_id: Optional[bytes32] = None
+
+    def to_program(self) -> Program:
+        condition: Program = Program.to([ConditionOpcode.AGG_SIG_PARENT, self.pubkey, self.msg])
+        return condition
+
+    @classmethod
+    def from_program(cls: Type[AggSigParent], program: Program, **kwargs: Any) -> AggSigParent:
+        return cls(
+            G1Element.from_bytes(program.at("rf").atom),
+            program.at("rrf").atom,
+            **kwargs,
+        )
```

Comment:
all of these classes have a lot of duplicated code. it seems like it should be possible to write a decorator to add the members.

Reply:
I think you're _probably_ right but the code duplicated is so small and there's just enough subtle differences across the classes that I'm not sure it's actually a win in terms of the effort to write such a decorator and the clarity of the code once it's been added.